pkg/eventfilter/filter_engine.go

```diff
@@ -0,0 +1,117 @@
+package eventfilter
+
+import (
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/types"
+	"github.com/dop251/goja"
+	"github.com/dop251/goja/ast"
+	"github.com/dop251/goja/parser"
+	"github.com/pkg/errors"
+)
+
+const timeout = time.Second * 2
+
+func ParseFilterExpr(src string) (*goja.Program, error) {
+	program, err := parser.ParseFile(nil, "", src, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := program.Body[0].(*ast.ExpressionStatement); !ok {
+		return nil, errors.WithStack(errors.New("program body should be just an expression: " + src))
+	}
+
+	return goja.CompileAST(program, false)
+}
+
+func RunFilter(event cloudevents.Event, program *goja.Program) (bool, error) {
+	vm := goja.New()
+	obj, err := configureEventObject(vm, event)
+	if err != nil {
+		return false, err
+	}
+	vm.Set("event", obj)
+
+	val, err := runWithSafeTimeout(timeout, vm, program)
+	if err != nil {
+		return false, err
+	}
+	return val.ToBoolean(), nil
+}
+
+func configureEventObject(vm *goja.Runtime, event cloudevents.Event) (*goja.Object, error) {
+	obj := vm.NewObject()
+	for name, fn := range map[string]func() interface{}{
+		"specversion": func() interface{} {
+			return event.SpecVersion()
+		},
+		"type": func() interface{} {
+			return event.Type()
+		},
+		"source": func() interface{} {
+			return event.Source()
+		},
+		"subject": func() interface{} {
+			return event.Subject()
+		},
+		"id": func() interface{} {
+			return event.ID()
+		},
+		"time": func() interface{} {
+			return event.Time()
+		},
+		"dataschema": func() interface{} {
+			return event.DataSchema()
+		},
+		"datacontenttype": func() interface{} {
+			return event.DataContentType()
+		},
+	} {
+		if err := bindValueExtractor(vm, name, fn, obj); err != nil {
+			return nil, errors.Wrap(err, "Error while binding event object")
+		}
+	}
+	for name, val := range event.Extensions() {
+		if err := bindValue(vm, name, val, obj); err != nil {
+			return nil, errors.Wrap(err, "Error while binding event object")
+		}
+	}
+	return obj, nil
+}
+
+func bindValue(vm *goja.Runtime, propName string, eventVal interface{}, obj *goja.Object) error {
+	return obj.DefineAccessorProperty(propName, vm.ToValue(func(fc goja.FunctionCall) goja.Value {
+		return coerceToJsTypes(vm, eventVal)
+	}), nil, goja.FLAG_FALSE, goja.FLAG_TRUE)
+}
+
+func bindValueExtractor(vm *goja.Runtime, propName string, extractFn func() interface{}, obj *goja.Object) error {
+	return obj.DefineAccessorProperty(propName, vm.ToValue(func(fc goja.FunctionCall) goja.Value {
+		return coerceToJsTypes(vm, extractFn())
+	}), nil, goja.FLAG_FALSE, goja.FLAG_TRUE)
+}
```

Comment:
They are similar, can we do
```golang
func bindValueExtractor(vm *goja.Runtime, propName string, extractFn func() interface{}, obj *goja.Object) error {
    return bindValue(vm, propName, extractFn(), obj)
}
```
Or get rid of `bindValueExtractor` completely?

Reply:
nope because i want to keep the extraction lazy, that's the reason why we have these two methods. in one case the `bindValueExtractor` enclose in the getter closure the function pointer, in the other case it enclose the value itself

Reply:
> nope because i want to keep the extraction lazy

What is the goal? Speed?

Reply:
> What is the goal? Speed?

Mostly avoid unnecessary memory allocations & type conversions to copy all the stuff inside the engine, when you just access a bunch of fields of the event